src/yellowpaper/chapters/03-implementation/data-disclosure.tex

```diff
@@ -1,64 +1,83 @@
 
-The protocol fairness is guaranteed by a contract on the public chain. The contract is able to hold money and is stateful: it is capable of storing a log $L$ with data. All the data that parties send to the contract are appended to $L$.
+\item \label{step:seller-deposits-money} If \tParty{E} agrees to proceed, she sends a predefined amount of money $C_E$ to the contract address. $C_E$ is a security deposit: if \tParty{E} tries to cheat, she would lose this money.
 
-\begin{enumerate}
-\item \tParty{E} creates a contract on the public chain and sends the following data to the contract: $\Sign_\Party{E}(C_D,\ \SizeOf(\Lock{D}), R, R_0)$. She also sends a predefined amount of money $C_E$ to the contract address. $C_E$ is a security deposit: if \tParty{E} tries to cheat, she would lose this money.
-\item The buyer generates a new keypair ($\PubKey{B},\ \SecretKey{B}$), and sends $C_D$ worth of money to the contract address. Along with the money, \tParty{B} sends the public key $\PubKey{B}$ of the newly generated keypair, and the size of the data.
-\item \tParty{E} transfers the encrypted data chunks $\Lock{E}$ to the buyer. \tParty{B} computes the Merkle root $R'$ and the size $s'$ of the received data $\Lock{E}'$:
+\item \tParty{E} transfers the encrypted data chunks $\Lock{D}$ directly to the buyer. \tParty{B} computes the Merkle root $R'$ and the size $s'$ of the received data $\Lock{D}'$:
 \begin{equation}
-R' = \Root(\MerkleTree(\Lock{E}'))
+R' = \Root(\MerkleTree(\Lock{D}'))
 \end{equation}
 \begin{equation}
-s' = \SizeOf(\Lock{E}')
+s' = \SizeOf(\Lock{D}')
 \end{equation}
-\item \tParty{B} makes a transaction with a receipt $\Sign_\Party{B}(\{R',\ s'\})$ to the contract address. The parties can proceed if and only if the following is true:
+\item \label{step:buyers-receipt} \tParty{B} makes a transaction with a receipt $\Sign_\Party{B}(\{R',\ s'\})$ to the contract address. The parties can proceed if and only if the following is true:
 \begin{equation}
 (R' = R)\ \land\ (s' = s)
 \end{equation}
 Otherwise, the protocol halts.
-\item \tParty{E} sends $\Sign_\Party{E}(\Encrypt_\Party{B}(k))$ to the contract.
-\item \tParty{B} decyphers and checks the received data. In case some data chunk $e_i \in \Lock{E}$ is invalid, \tParty{B}~sends a transaction with~$\{\SecretKey{B},\ e_i,\ \Path(e_i,\ \MerkleTree(\Lock{E})),\ \Path(d_i,\ \MerkleTree(\Lock{D}))\}$ to the contract. By doing so, \tParty{B}~reveals the data chunk~$d_i$ corresponding to the encrypted chunk~$e_i$.  She also shares proof that~$e_i$ was indeed part of a Merkle tree with root~$R$ and the~$d_i$ was part of a Merkle tree~$R_0$. The contract checks the validity of $d_i$ and decides whether \tParty{B}~has rightfully accused~\tParty{E} of cheating.
```

Comment:
Wait, _secret key of `B`_ is sent to contract? Shouldn't this be the `k` used to turn `D` into `D_locked`?

Reply:
No. If we initially transfer `k` off-chain, the Buyer can get data for free: he should just wait for timeout and pretend that the Seller has never transferred the data. Since the Seller transfers `k` on-chain he has to encrypt it (NB: could you remind me, why?).

When the Buyer reveals `(sk_B, e_i)`, the witnesses decipher `k` and then get `{ d_i = (T_priv, P_priv, H, P_pub) }`.

Reply:
>NB: could you remind me, why?
To protect against eavesdroppers, collecting encrypted off-chain data via network tools, ofk.